extensions/kinesis/src/main/java/com/hazelcast/jet/kinesis/impl/KinesisSourceP.java

```diff
@@ -0,0 +1,185 @@
+        this.eventTimeMapper = new EventTimeMapper<>(eventTimePolicy);
+        this.hashRange = Objects.requireNonNull(hashRange, "hashRange");
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) throws Exception {
+        super.init(context);
+
+        logger = context.logger();
+        id = context.globalProcessorIndex();
+
+        helper = new KinesisHelper(kinesis, stream, logger);
+
+        logger.info("Processor " + id + " handles " + hashRange);
+
+        helper.waitForStreamToActivate();
+        List<Shard> shardsInRange = helper.listShards(
+                (Predicate<? super Shard>) shard -> shardBelongsToRange(shard, hashRange));
+        rangeMonitor = new RangeMonitor(context.totalParallelism(), kinesis, stream, hashRange, shardsInRange, logger);
+        addShardReaders(shardsInRange);
+    }
+
+    @Override
+    public boolean complete() {
+        if (!emitFromTraverser(traverser)) {
+            return false;
+        }
+
+        runMonitor();
+        runReaders();
+        return false;
+    }
+
+    private void runMonitor() {
+        RangeMonitor.Result result = rangeMonitor.run();
+        if (RangeMonitor.Result.NEW_SHARDS.equals(result)) {
+            Collection<Shard> shards = rangeMonitor.getNewShards();
+            addShardReaders(shards);
+        }
+    }
+
+    private void runReaders() {
+        for (int i = 0; i < shardReaders.size(); i++) {
+            int currentReader = nextReader;
+            ShardReader reader = shardReaders.get(currentReader);
+            nextReader = incrCircular(currentReader, shardReaders.size());
+
+            ShardReader.Result result = reader.run();
+            if (ShardReader.Result.HAS_DATA.equals(result)) {
+                Record[] records = reader.getData();
+                traverser = Traversers.traverseArray(records)
+                        .flatMap(record -> eventTimeMapper.flatMapEvent(
+                                entry(record.getPartitionKey(), record.getData().array()), //todo: shady?
+                                currentReader,
+                                record.getApproximateArrivalTimestamp().getTime()
+                        ));
+                emitFromTraverser(traverser);
+                return;
+            } else if (ShardReader.Result.CLOSED.equals(result)) {
+                Shard shard = reader.getShard();
+                logger.info("Shard " + shard.getShardId() + " of stream " + stream + " closed");
+                removeShardReader(currentReader);
+                nextReader = 0;
+                return;
+            }
+        }
+
+        traverser = eventTimeMapper.flatMapIdle();
+        emitFromTraverser(traverser);
+    }
+
+    @Override
+    public boolean saveToSnapshot() {
+        if (!emitFromTraverser(traverser)) {
+            return false;
+        }
+
+        //todo: actual snapshot saving; we will be saving the sequence numbers of last seen messages, per shard
```

Comment:
I think we should also save the list of closed shards until we're sure they will be no longer returned from `ListShards`.

Reply:
Ok, will keep in mind, have set a todo.